### DISCUSSION TITLE
Allow calculated and dynamic default values

### DISCUSSION THREAD

**Discussion by kgodey** [mathesar-foundation/mathesar#2272] [Feature Ideas]
Mathesar should support adding calculated default values for columns. Users would set up a formula for how the default value is calculated.

The calculated default values should be able to supoort both formulas based on existing columns, and dynamic values (eg., 'today' for date columns).

See some preliminary research and design work is published in the following wiki page:

- [Dynamic Defaults Proto-Spec](https://wiki.mathesar.org/product/specs/dynamic-defaults-proto/)

Related features:

- [Support for calculated columns in tables](https://github.com/mathesar-foundation/mathesar/discussions/2271)
- [Support for calculated columns to the Data Explorer](https://github.com/mathesar-foundation/mathesar/discussions/2274)

---

*If you'd like us to build this feature, please comment below with more details about how you plan to use this functionality.*

**Comment by jeromeperez** [comment]
Hi 
It would be interesting provided you could also use the columns of different tables.
But the link https://wiki.mathesar.org/en/design/specs does'nt function

**Reply by seancolsen** [reply]
@jeromeperez

Sorry about that broken link. I've edited the description above to link to two relevant wiki pages — though I'll add the caveat that these links don't offer much insight into the direction we're likely to head with this sort of feature. That's in fact because we've not yet made very many decisions about how we'll eventually end up implementing calculated and dynamic default values.

> It would be interesting provided you could also use the columns of different tables.

This sort of feedback is really helpful! Would you be interested in expanding on this idea by providing an example? What would the tables look like and what would the calculated values look like?

**Reply by jeromeperez** [reply]
first exemple in the same table :
to transform duration (format time 07:45) in a calculated field Nb_hour (7,75) with formula (duration - int(duration)) * 24

**Reply by jeromeperez** [reply]
After further study, when it concerns several tables, it's no longer a formula but rather a calculation based on the result of a query, so I don't know if it would be possible to integrate this into mathesar.
exemple:
table taxa fields totalfragment, taxa
table sample fields fragment, sample
table sample_taxa fields taxa,sample
totalfragment = Sum(fragment) with sample_taxa.taxa = taxa and sample_taxa.sample = sample

**Reply by seancolsen** [reply]
Thanks @jeromeperez. I also edited the description above to link to some other feature discussions:

- [Support for calculated columns in tables](https://github.com/mathesar-foundation/mathesar/discussions/2271)
- [Support for calculated columns to the Data Explorer](https://github.com/mathesar-foundation/mathesar/discussions/2274)

I think those might actually be more relevant the features you're describing here. Your first example would be calculated columns in tables. And for a calculation based on the result of a query, we're hoping that building this sort of functionality into the Data Explorer would allow you to define complex computations using data across multiple tables.

**Comment by jeromeperez** [comment]
> Thanks @jeromeperez. I also edited the description above to link to some other feature discussions:
> 
> * [Support for calculated columns in tables](https://github.com/mathesar-foundation/mathesar/discussions/2271)
i add my first example to this feature
> * [Support for calculated columns to the Data Explorer](https://github.com/mathesar-foundation/mathesar/discussions/2274)
and my second here ;-)
> 
> I think those might actually be more relevant the features you're describing here. Your first example would be calculated columns in tables. And for a calculation based on the result of a query, we're hoping that building this sort of functionality into the Data Explorer would allow you to define complex computations using data across multiple tables.

**Comment by nozense** [comment]
This feature is the first one i missed when trying Mathesar! This is really needed!
But does not [issues/2754](https://github.com/mathesar-foundation/mathesar/issues/2754) imply that this is already planed, and high priority? - if so it would help someone like me, looking over "Feature Ideas" to see if what I'd like to see is already in here, if it was marked as planned or similar!
